unittests/sha256: separated tests to prevent stack corruption on boards

BytesGalore opened the pull request:
Rationale:
Separating the individual testcases is necessary to balance the ram usage on boards.
Without this PR the sha256 unittests won't run for instance on the samr21-xpro board and constantly hardfault it.

OlegHahm commented:
I merged this PR, but the tests-hashes unittests still is constantly hard-faulting.

BytesGalore commented:
hmm strange, I can run the tests of this PR on a samr21-xpro board successfully 
`OK (10 tests)`.
I will also separate the initial sha256 hashes computation tests individually, maybe this helps additionally.

BytesGalore commented:
Did that, now its `OK (18 tests)` on my samr21-xpro.
But its not for free,
before split all tests to individual test-cases:
```
   text	   data	    bss	    dec	    hex	filename
  20852	    196	   3100	  24148	   5e54	RIOT/tests/unittests/bin/samr21-xpro/unittests.elf
```
and after separating them:
```
   text	   data	    bss	    dec	    hex	filename
  21368	    196	   3100	  24664	   6058	RIOT/tests/unittests/bin/samr21-xpro/unittests.elf
```

OlegHahm commented:
How did you compile and flash?

BytesGalore commented:
`make BOARD=samr21-xpro tests-hashes`
`make BOARD=samr21-xpro tests-hashes flash`
that's it.

OlegHahm commented:
Strange, I still get
```
2016-02-23 12:05:47,536 - INFO # �main(): This is RIOT! (Version: 2016.03-devel-545-g4cf50-tbilisi-sha256_separate_unittests)
2016-02-23 12:05:47,553 - INFO # ...........*** RIOT kernel panic:
2016-02-23 12:05:47,555 - INFO # HARD FAULT HANDLER
2016-02-23 12:05:47,555 - INFO # 
2016-02-23 12:05:47,556 - INFO # *** rebooting...
2016-02-23 12:05:47,570 - INFO # �main(): This is RIOT! (Version: 2016.03-devel-545-g4cf50-tbilisi-sha256_separate_unittests)
2016-02-23 12:05:47,589 - INFO # ...........*** RIOT kernel panic:
2016-02-23 12:05:47,589 - INFO # HARD FAULT HANDLER
2016-02-23 12:05:47,589 - INFO # 
2016-02-23 12:05:47,591 - INFO # *** rebooting...
```
Which toolchain are you using?

BytesGalore commented:
`gcc version 4.9.3 20150303 (release) [ARM/embedded-4_9-branch revision 221220] (GNU Tools for ARM Embedded Processors)`

BytesGalore commented:
Its the 2015q1 version

OlegHahm commented:
Hm, interesting, with 4.9.2 it's succeeding here as well. With 5.3 it's failing.

OlegHahm commented:
Apparently all gcc versions >= 5 are failing.

BytesGalore commented:
ok. Do other unittests succeed with >= 5?

OlegHahm commented:
Partly. I checked `tests-fib` -> success and `tests-crc` -> error, so far.

OlegHahm commented:
It seems still related to `test_hashes_sha256_hash_long_sequence()`.

BytesGalore commented:
I had to download the recent arm 5 toolchain (my internet connection is quite ancient). 
Now I use `gcc version 5.2.1 20151202 (release) [ARM/embedded-5-branch revision 231848]`
and can reproduce the behaviour.
With the >=5 gcc, the codeimage size is way larger with  for this PR 
in dec. 
4.9.2: 21368 
5.2.1: 24184

OlegHahm commented:
That's interesting, too. With `arm-none-eabi-gcc (Arch Repository) 5.3.0`, I have 20644 for ROM. Or are you referring to the overall size? Then it's 23940 here.

BytesGalore commented:
overall (the output given when the build-process finishes)

OlegHahm commented:
So, here it is
```
   text	   data	    bss	    dec	    hex	filename
  20644	    196	   3100	  23940	   5d84	/home/oleg/git/RIOT/tests/unittests/bin/samr21-xpro/unittests.elf
```
for the above mentioned toolchain.

BytesGalore commented:
Funny (not really), with the mentioned toolchain I get:
```
   text	   data	    bss	    dec	    hex	filename
  20888	    196	   3100	  24184	   5e78	RIOT/tests/unittests/bin/samr21-xpro/unittests.elf
```
removed `RIOT/tests/unittests/bin/` folder and then built

BytesGalore commented:
(I mean my toolchain)

BytesGalore commented:
Oh, I see now I mixed up the sizes

BytesGalore commented:
They become smaller with each version

BytesGalore commented:
I removed `test_hashes_sha256_hash_long_sequence()`

BytesGalore commented:
I checked all unittests with `gcc version 5.2.1 20151202 (release) [ARM/embedded-5-branch revision 231848]`.
Only `tests-cbor` results in hardfault the samr21-xpro, but it also does with the 4.9.3 toolchain

OlegHahm commented:
Ah, yes, sorry, I also tried `tests-cbor` not `tests-crc`. Hm, now we have to find the similarities between the cbor and the sha256 unit tests.

OlegHahm commented:
> I removed test_hashes_sha256_hash_long_sequence()

I confirmed that this is working now, but have you tried if by removing this test the unittest wouldn't work on SAMR21 even without this PR?

BytesGalore commented:
I tried something different, I removed all other tests but `test_hashes_sha256_hash_long_sequence()` and it worked without hardfault.

OlegHahm commented:
Have you tried to tweak the stack sizes?

BytesGalore commented:
I tried it as you suggested, just removed the specific test with the long string (on master without this PR) but still hardfault. Removing additional tests makes the hardfault go away.

I didn't try to stretch the stacksize yet, I hoped to get the stackusage below its border by reorganizing the unittests.
And it works to some extent, but seems to have not enough room for the long test.

BytesGalore commented:
I doubled the stacksize `THREAD_STACKSIZE_DEFAULT` to 2048 and the initial `tests-hashes` does not hardfault on the samr21-xpro.
But it doesn't seem to help in this PR, I tried several values between 1024 and 4096 without success.

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+
+static void test_hashes_hmac_sha256_hash_sequence(void)
+{
+    unsigned char key[64];
```
make this var `static`, the function will not be thread safe, but I don't think it is going to be called from multiple threads simultaneously, and there is a static buffer below as well so there's that too..

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+static void test_hashes_hmac_sha256_hash_PRF6(void)
+{
+    /* Test Case PRF-6: */
+    const char strPRF6[] = "This is a test using a larger than block-size key and a "
```
@BytesGalore try making this and all other strings `static const` to see if it has any effect on your issues on samr21-xpro

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+    const char strPRF6[] = "This is a test using a larger than block-size key and a "
+                           "larger than block-size data. The key needs to be hashed "
+                           "before being used by the HMAC algorithm.";
+    unsigned char longKey[131];
```
try `static` here too

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+                           "before being used by the HMAC algorithm.";
+    unsigned char longKey[131];
+    static unsigned char hmac[SHA256_DIGEST_LENGTH];
+    memset(longKey, 0xaa, 131);
```
`memset(longKey, 0xaa, sizeof(longKey));`

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+    memset(longKey, 0xaa, 131);
+
+    /* the same key is used as above: 131 x 0xa */
+    hmac_sha256(longKey, 131, (unsigned*)strPRF6, strlen(strPRF6), hmac);
```
sizeof()

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+{
+    /* Test Case PRF-5: */
+    const char strPRF5[] = "Test Using Larger Than Block-Size Key - Hash Key First";
+    unsigned char longKey[131];
```
static

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+    const char strPRF5[] = "Test Using Larger Than Block-Size Key - Hash Key First";
+    unsigned char longKey[131];
+    static unsigned char hmac[SHA256_DIGEST_LENGTH];
+    memset(longKey, 0xaa, 131);
```
sizeof

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+    static unsigned char hmac[SHA256_DIGEST_LENGTH];
+    memset(longKey, 0xaa, 131);
+
+    hmac_sha256(longKey, 131, (unsigned*)strPRF5, strlen(strPRF5), hmac);
```
sizeof

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+    /*
+    * set key to: 0102030405060708090a0b0c0d0e0f10111213141516171819
+    */
+    for (size_t i = 0; i < 25; ++i) {
```
sizeof

gebart commented on tests/unittests/tests-hashes/tests-hashes-sha256-hmac.c:
```diff
@@ -0,0 +1,172 @@
+{
+    /* Test Case PRF-4: */
+    char strPRF4[50];
+    unsigned char key[25];
```
try static

BytesGalore commented:
@gebart I tested it using `static` for all strings, but unfortunately it didn't help

gebart commented:
Try allocating a global buffer instead of keeping one in each test case, it might save a bit of RAM

BytesGalore commented:
I converted the expected hash strings to global byte arrays, which saves about 1/2 of used RAM for them. In conjunction I also removed `compare_str_vs_digest()` which also saved a fistful of bytes on the stack.
Eventually the squeezing was successful. 
Now the unittest doesn't produce a hardfault on the samr21-xpro (including the long sequence test), and all tests pass :D.

OlegHahm commented:
Really takes some time on SAMR21, but it seems to work. Great job! ACK. Ready to squash.

BytesGalore commented:
squashed

BytesGalore commented:
nice, thx for the review :)
